### DISCUSSION TITLE
Intl.NumberFormat hydration mismatch

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [nuxt/nuxt#17629] [Questions]
When using `new Intl.NumberFormat`, the following error is printed in the browser console on page load:
```bash
[Vue warn]: Hydration text mismatch:
- Client: "22,00 $US"
- Server: "US$22.00" 
  at <[slug] onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > key="/products/t-shirt" > 
  at <BaseTransition mode="out-in" appear=false persisted=false  ... > 
  at <Transition name="page" mode="out-in" > 
  at <RouterView > 
  at <NuxtPage> 
  at <Index> 
  at <App key=1 > 
  at <NuxtRoot>
```

I have been searching for a few hours now and I think I identified the causes of the issue. The locales in browser and Node are not the same and so, if not included in Node for example, the `en` locale will be picked by default.
So I updated my Node version to `16.15.0` but I'm still getting the error.
I also saw (and tried implementing) the `@formatjs/vue-intl` package but couldn't get it to work with `iso2` country codes.

How should I do?

Right now, I use the `ClientOnly` but it feels like a dirty hack...

**Comment by tommie** [comment]
I have the same issue, but for DateTimeFormat. Actually two problems:

1. The locale might differ between SSR and client, and
2. the timezone might also differ.

I don't have a complete solution yet.

On first access, SSR uses whatever the default locale is on the backend. I extract locale information using `Intl.DateTimeFormat.resolvedOptions()` on the client, and set a cookie with that information, so the next time SSR happens, it can pick up the cookie and use that instead of its default. (I also pass the cookie value in a `useState` to ensure consistency, though it shouldn't be needed.) This handles both locale and timezone.

However, the implementations might still be different, so there will be the same hydration mismatch as you have/had. Since I'd prefer doing this formatting as late as possible (in a component,) I can't easily use `useState` to pass the data from SSR; I have no unique ID for an individual component that I can use as the state key.

A dev-friendly solution would be to incorporate `<FormattedDateTime>` and `<FormattedNumber>` into Nuxt, so the runtime engine could check the inputs for mismatches, rather than the rendered result.

A partial solution is to derive the state key from the input `Date` and formatting options, essentially deduplicating states. This would ensure no hydration mismatch, but would of course entangle reactivity in the client. The adding a `watch` to do the same formatting reactively. This is emulating a `computed` with a de-duplicating `useState` in the middle (untested code):

```ts
function formatDate(value?: Date) {
  return value ? new Intl.DateTimeFormat(locales.value, options.value).format(value) : "";
}
    
const formattedTime = useState(
  `${props.value}/${JSON.stringify(locales.value)}/${JSON.stringify(options.value)}`,
  () => formatDate(props.value)
);

watch(() => props.value, (value) => {
  formattedTime.value = formatDate(value);
}, { immediate: true });
```
